databases/nmdc/database.go

```diff
@@ -0,0 +1,992 @@
+// Copyright (c) 2023 The KBase Project and its Contributors
+// Copyright (c) 2023 Cohere Consulting, LLC
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy of
+// this software and associated documentation files (the "Software"), to deal in
+// the Software without restriction, including without limitation the rights to
+// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
+// of the Software, and to permit persons to whom the Software is furnished to do
+// so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package nmdc
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"net/url"
+	"os"
+	"slices"
+	"strconv"
+	"strings"
+	"time"
+	"unicode"
+
+	"github.com/google/uuid"
+
+	"github.com/kbase/dts/config"
+	"github.com/kbase/dts/credit"
+	"github.com/kbase/dts/databases"
+	"github.com/kbase/dts/frictionless"
+)
+
+// file database appropriate for handling searches and transfers
+// (implements the databases.Database interface)
+type Database struct {
+	// database identifier
+	Id string
+	// ORCID identifier for database proxy
+	Orcid string
+	// HTTP client that caches queries
+	Client http.Client
+	// authorization info
+	Auth authorization
+	// mapping of host URLs to endpoints
+	EndpointForHost map[string]string
+}
+
+func NewDatabase(orcid string) (databases.Database, error) {
+	if orcid == "" {
+		return nil, databases.UnauthorizedError{
+			Database: "nmdc",
+			Message:  "No ORCID ID was given",
+		}
+	}
+
+	nmdcUser, haveNmdcUser := os.LookupEnv("DTS_NMDC_USER")
+	if !haveNmdcUser {
+		return nil, databases.UnauthorizedError{
+			Database: "nmdc",
+			Message:  "No NMDC user (DTS_NMDC_USER) was provided for authentication",
+		}
+	}
+	nmdcPassword, haveNmdcPassword := os.LookupEnv("DTS_NMDC_PASSWORD")
+	if !haveNmdcPassword {
+		return nil, databases.UnauthorizedError{
+			Database: "nmdc",
+			Message:  "No NMDC password (DTS_NMDC_PASSWORD) was provided for authentication",
+		}
+	}
+
+	// check for "nersc" and "emsl" Globus endpoints
```

Comment:
this comment might make more sense above the next stanza (`for _, functionalName ...`) - you can then remove the comment saying "was this functional name assigned..."